### ISSUE TITLE
Idea: use a search engine to power search

### ISSUE THREAD

**Issue by vignesh-sankaran** [rust-lang/crates.io#1041]
Just putting an idea out there, maybe we could look into using something like [Lucene](https://lucene.apache.org/core/) or ElasticSearch (powered by Lucene too if I'm not mistaken) sometime down the road. I interned in the Search Quality team at Seek Limited a couple of years back, so can ask some developers about some of the intricacies of setting this up if this is something of interest :).

**Comment by kureuil** [comment]
An [https://www.algolia.com/](Algolia "Algolia's website") employee once [said](https://github.com/rust-lang/crates.io/issues/205#issuecomment-153672545) that we were eligible for their free plan.
It could be simpler to setup in the near-term, though it's probably "safer" to be in control of all of the stack (also would avoid private crates.io instances depending on an external service).

**Comment by vignesh-sankaran** [comment]
Yep, I feel we should be running it locally, because if we change the way we do search to an online service, it could be a problem for a private crates.io instance running behind a corporate firewall. That and we can't develop + test crates.io offline :).

So I've come back with a bag of alternative Rust-based search engines too we could look at, depending on what features we need and how complete they are. If this is something of interest to implement, I can start researching how to set this up. If I'm going to work on this, it could be a while before I get to this.

- [tantivy](https://github.com/tantivy-search/tantivy)
- [noise](https://github.com/pipedown/noise)
- [perlin](https://github.com/CurrySoftware/perlin)
- [kite](https://github.com/kaedroho/kite)
- [rusticsearch](https://github.com/kaedroho/rusticsearch)

**Comment by carols10cents** [comment]
I'm interested in knowing exactly how a different search will improve on results, and what they make possible that isn't possible with postgres, before adding another dependency. Postgres has quite powerful search capabilities, but we're definitely not using them all in the best ways we could be. I'd be worried we'd switch to something else that we also wouldn't spend time tuning and be no better off except for an extra dependency.

This may involve researching the different systems, and it may involve actually using them and comparing results to what we have today.

**Comment by vignesh-sankaran** [comment]
This is something that needs to be researched, and I'm currently unable to do that :(

**Comment by Deedasmi** [comment]
Just tagging #370 on here as an alternative and #159 as a motivation.

**Comment by erlend-sh** [comment]
I'd love to chat more about this if there's still interest. We made a live demo for Crates powered by MeiliSearch:

https://crates.meilisearch.com/

As a first trial project, you could start by using MeiliSearch just for its client-side, search-as-you-type functionality. You could self-host MeiliSearch or we'd provide you with managed API hosting for free.

**Comment by TheGP** [comment]
The https://crates.meilisearch.com/ even shows proper results for "dotenv" query, I'm switching to it, thanks @erlend-sh!

So will there be any movements regarding improving the search? It should be a priority as finding the right tools for building a project is important. You don't have resources or are you working on some less important tasks?
